dask/array/routines.py

```diff
@@ -689,53 +692,117 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
             "See the numpy.histogram docstring for more information."
         )
 
+    dependencies = [a]
+
+    range_refs = None
+    if range is not None:
+        try:
+            if len(range) != 2:
+                raise ValueError(
+                    f"range must be a sequence of length 2, but got {len(range)} items"
+                )
+        except TypeError:
+            raise TypeError(f"expected a sequence for range, not {range}") from None
+
+        range_refs = []
+        for elem in range:
+            if isinstance(elem, Array):
+                if elem.shape != ():
+                    raise ValueError(
+                        f"Dask arrays passed in the range argument to histogram must be scalars "
+                        f"(shape of `()`); got one with shape {elem.shape}."
+                    )
+                dependencies.append(elem)
+                key = elem.__dask_keys__()[0]
+                range_refs.append(key)
+            else:
+                range_refs.append(elem)
+
     if not np.iterable(bins):
+        assert not isinstance(
+            bins, Array
+        ), "The number of bins cannot be a Dask array, it must be an actual value"
         bin_token = bins
-        mn, mx = range
-        if mn == mx:
-            mn -= 0.5
-            mx += 0.5
+        if len(dependencies) == 1:
+            # ^ `len(dependencies) == 1` here iff neither element in `range` was a Dask array.
+            # `mn == mx` could be an expensive implicit compute otherwise.
+            mn, mx = range
+            if mn == mx:
+                mn = mn - 0.5
+                mx = mx + 0.5
+
+            bins = np.linspace(mn, mx, bins + 1, endpoint=True)
+        else:
+            linspace_name = tokenize(bins, *range_refs)
+            linspace_dsk = {
+                (linspace_name, 0): (
+                    np.linspace,
+                    range_refs[0],
+                    range_refs[1],
+                    bins + 1,
+                )
+            }
+            linspace_graph = HighLevelGraph.from_collections(
+                linspace_name, linspace_dsk, dependencies=dependencies[1:]
+            )
+            # ^ TODO: dask linspace doesn't support delayed values
+            bins = Array(linspace_graph, linspace_name, [(bins + 1,)], dtype=float)
 
-        bins = np.linspace(mn, mx, bins + 1, endpoint=True)
     else:
         bin_token = bins
+
     token = tokenize(a, bin_token, range, weights, density)
 
+    if isinstance(bins, Array):
+        if bins.ndim != 1:
+            raise ValueError(f"bins must have 1 dimension, got {bins.ndim}")
+        dependencies.append(bins)
+        bins_ref = "histogram-bins-" + token
+
+        bins_dsk = {bins_ref: (np.concatenate, bins.__dask_keys__())}
```

Comment:
I recommend using `dask.delayed.unpack_collections` here instead of generating your own task. Something like:

```python
bins_task, deps = unpack_collections(bins)
bins_graph = HighLevelGraph.from_collections(
    bins_ref, {bins_ref: bins_task}, dependencies=deps
)
```

Reply:
I make this change; there are no issues around this import if you've installed `dask[array]` but not `dask[delayed]`, right? Just want to double check.

Reply:
array depends on delayed, so that's fine.